### ISSUE TITLE
Type class to serialise private keys

### ISSUE THREAD

**Issue by geigerzaehler** [oscoin/oscoin#664]
## Problem

For neither crypto implementation `c` does `PrivateKey c` have an instance of `Serialise`. This is for good reason since we want to avoid leaking private keys. Still we need to serialize private keys to store them on disk. For this purpose we have `Oscoin.Crypto.PubKey.RealWorld.serialisePrivateKey` and `Oscoin.Crypto.PubKey.FileStore`. This code is however monomorphic in the crypto used. This means that any code using this must also be monomorphic.

## Solution/implementation plan

Let’s create a `SerialisePrivateKey` type class that is similar to `Serialise` with methods `serialisePrivateKey` and `deserialisePrivateKey`. This will enable us to make the `FileStore` module polymorphic with the `SerialisePrivateKey` constraint. It will also prevent us from using the private key with code that requires the `Serialise` constraint and might leak the secret key.

##  Alternative solutions considered

* We could stick with monomorphic code.
* We could pass the serialization and deserialization functions explicitly.

**Comment by geigerzaehler** [comment]
@kim, If I remember correctly you initially pointed out that we should not implement `Serialise` for `PrivateKey` because of the security risk so I’d like your input here.

**Comment by kim** [comment]
It sounds reasonable to create a typeclass for that, but I'd expect the _storage_ to determine ther format, not the crypto used. For example, you might want to create some backend which takes `.pem` encoded keys, or a GPG backend which takes... whatever. So perhaps it would be more flexible to define a class to get at the raw bytes (`ScrubbedBytes`, possibly) of the keypair, and leave it to the store to decide what to do with them.

**Comment by geigerzaehler** [comment]
Yes, that makes sense. How do you feel about adding methods `privateKeyBytes :: PrivateKey c -> ScrubbedBytes` and `privateKeyFromBytes :: ScrubbedBytes -> Maybe (PrivateKey c)` to the `PubKey` type class instead of using another type class?

**Comment by kim** [comment]
> instead of using another type class

Hm, yeah, not sure. It would probably make the interface (more) confusing, as the corresponding facilities for public keys are implemented as instances of some other classes.

**Comment by geigerzaehler** [comment]
> > instead of using another type class
> 
> Hm, yeah, not sure. It would probably make the interface (more) confusing, as the corresponding facilities for public keys are implemented as instances of some other classes

Makes sense. I updated the description accordingly.
